### PR TITLE
Real-time build log stream

### PR DESCRIPTION
**Pull request by KrauseFx**
Fixes https://github.com/fastlane/ci/issues/201
Fixes https://github.com/fastlane/ci/issues/147

**To test this change:**
1. open a project view (e.g. `http://localhost:9292/projects/details/db799377-aaa3-4605-ba43-c91a13c8f83`)
1. Click on `Trigger job now`
1. See how half the build output is there instantly (shows that fastlane.ci "caches" the output)
1. See how more messages are being loaded as _fastlane_ runs the `Fastfile` (has a few sleeps in there to proof the real-time-ness)

**Notes**
1. Thin seems to require the use of `localhost` instead of `127.0.0.1`, I didn't look into this yet if we can support both
1. Main thing that's missing to get this in a rather solid state is the identification of a build. As raised the question in Slack (`I see the Build object doesn't have an ID like we have for all other objects. Is there a reason for that? Did we decide you can't run multiple builds for one sha? Is project + sha unique?)`). My proposal is to add a unique ID to builds also, so that this ID in combination with the project is unique (or even without the project?). Once that's done, multiple `# TODO` items from this PR can be resolved. (e.g. see https://github.com/fastlane/ci/pull/200/files#diff-40d0c75f55437cd62b7224bd7d90d275R14)
1. The custom middleware holds a reference to each web socket connection (https://github.com/fastlane/ci/pull/200/files#diff-7db6fa3652598445f8cb2adaed28ae88R40) grouped by the Build ID, so once the build ID is unique, without the project ideally, this will work nicely.
1. The frontend is just a basic proof of concept of course
1. We discussed that we don't want to generate the HTML on the backend long term. I'll change this with a follow-up PR
1. The `fastlane.log` file is hard coded for now, until we have the artifact system ready. 
1. I get those weird `lib/git/lib.rb:460: warning: conflicting chdir during another chdir block` warnings again. I assume this happens due to running _fastlane_ and we do some `Dir.chdir` in that code base also. I haven't verified yet if this PR adds this issue, or if that's already the case in `master`.
1. As seen in https://github.com/fastlane/ci/pull/200/files#diff-62e64a60028b7cd0fdf6248bbc3f9bfcR13, we currently have to store references to all active `test_runner_services` to keep them alive, and to be able to access them when the user wants to see their progress. I added a comment with a proposal on how we can properly architect it

### CODE REVIEW

File: services/test_runner_service.rb [hunk 1]
```diff
@@ -11,14 +27,57 @@ class TestRunnerService
     attr_accessor :current_build
     attr_accessor :sha
 
-    def initialize(project: nil, sha: nil, github_service: nil)
+    # All lines that were generated so far, this might not be a complete run
+    # This is an array of hashes
+    # TODO: have a class representing a Row (has to offer dynamic values though, as we might have non fastlane runners in the future)
+    attr_accessor :all_lines
```
Comment (taquitos) [inline services/test_runner_service.rb]: `all_build_output_log_lines` might be clearer

File: services/test_runner_service.rb [hunk 2]
```diff
@@ -11,14 +27,57 @@ class TestRunnerService
...
+
+    # All blocks listening to changes for this build
+    attr_accessor :all_blocks
```
Comment (taquitos) [inline services/test_runner_service.rb]: `build_change_observer_blocks`

File: features/build/build_controller.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+module FastlaneCI
+  # Controller for a single project view. Responsible for updates, triggering builds, and displaying project info
+  class BuildController < AuthenticatedControllerBase
+    HOME = "/projects/*/builds"
+
+    use(FastlaneCI::BuildWebsocketBackend)
+
+    get "#{HOME}/*" do |project_id, build_id|
+      project = self.user_project_with_id(project_id: project_id)
+      build = project.builds.find { |b| b.sha == build_id } # TODO: We need a build ID, sha isn't enough
```
Comment (minuscorp) [inline features/build/build_controller.rb]: This should be just the number?

File: features/test_runner/fastlane_test_runner.rb [hunk 4]
```diff
@@ -4,13 +4,35 @@ module FastlaneCI
...
+          #
+          # {:type=>:success, :message=>"Everything worked"}
+          #
+          # we append the HTML code that should be used in the `html` key
+          # the result looks like this
+          #
+          # {"type":"success","message":"Driving the lane 'ios beta'","html":"<p class=\"success\">Driving the lane 'ios beta'</p>"}
+          #
+          row[:html] = FastlaneOutputToHtml.convert_row(row)
           yield(row)
```
Comment (minuscorp) [inline features/test_runner/fastlane_test_runner.rb]: This was failing for me (on `master`) `yield` misses a callable, where is it? 😕

Comment (KrauseFx) [reply features/test_runner/fastlane_test_runner.rb]: Sorry, my bad that it was failing on master, it's over here now https://github.com/fastlane/ci/blob/e1edb77c5904036f8b8ba6280fadc04dd0df9242/services/test_runner_service.rb#L108-L110

File: features/build/build_controller.rb (hunk 3 above)
Comment (KrauseFx) [reply features/build/build_controller.rb]: Yes, once we resolved the question/proposal around having a unique Build ID

**Comment by minuscorp** [comment]
Can we get this merged? So I can follow with #205 😄

**Comment by KrauseFx** [comment]
Sorry for the delay @minuscorp, this is merged now 👍

**Pull request merged**